src/OpenTelemetry.Exporter.Jaeger/Implementation/EnumerationHelper.cs

```diff
@@ -0,0 +1,197 @@
+﻿// <auto-generated>
+// <copyright file="EnumerationHelper.cs" company="OpenTelemetry Authors">
+// Copyright 2018, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Collections;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Reflection;
+using System.Reflection.Emit;
+
+namespace OpenTelemetry.Exporter.Jaeger.Implementation
+{
+    internal class DictionaryEnumerator<TKey, TValue, TState> : Enumerator
+        <IEnumerable<KeyValuePair<TKey, TValue>>,
+        KeyValuePair<TKey, TValue>,
+        TState>
+        where TState : struct
+    {
+        protected DictionaryEnumerator()
+        {
+        }
+    }
+
+    internal class ListEnumerator<TValue, TState> : Enumerator
+        <IEnumerable<TValue>,
+        TValue,
+        TState>
+        where TState : struct
+    {
+        protected ListEnumerator()
+        {
+        }
+    }
+
+    // A helper class for enumerating over IEnumerable<TItem> without allocation if a struct enumerator is available.
```

Comment:
is there a feature proposal on runtime to support it natively? It would be great to reference it here.

Reply:
> Saves on the allocations but it's kind of dark magic?

It does, at the expense of a whole bunch of start-up overhead and complexity as well as unorthodox consumption patterns.  It's certainly neat and if it's the right trade-off for this project, great, but I don't think it makes sense as a general-purpose feature in dotnet/runtime.